execution_engine/src/core/engine_state/upgrade.rs

```diff
@@ -328,4 +342,70 @@ where
 
         Ok(())
     }
+
+    pub(crate) fn ensure_consistent_contract_state(
+        &self,
+        correlation_id: CorrelationId,
+        handle_payment_hash: &ContractHash,
+        engine_config: &EngineConfig,
+    ) -> Result<(), ProtocolUpgradeError> {
+        let mut address_generator = {
+            let seed_bytes = (self.old_protocol_version, self.new_protocol_version).to_bytes()?;
+
+            let phase = Phase::System;
+
+            AddressGenerator::new(&seed_bytes, phase)
+        };
+
+        let system_contract = SystemContractType::HandlePayment;
+        let contract_name = system_contract.contract_name();
+        let mut contract = if let StoredValue::Contract(contract) = self
+            .tracking_copy
+            .borrow_mut()
+            .read(correlation_id, &Key::Hash(handle_payment_hash.value()))
+            .map_err(|_| {
+                ProtocolUpgradeError::UnableToRetrieveSystemContract(contract_name.to_string())
+            })?
+            .ok_or_else(|| {
+                ProtocolUpgradeError::UnableToRetrieveSystemContract(contract_name.to_string())
+            })? {
+            contract
+        } else {
+            return Err(ProtocolUpgradeError::UnableToRetrieveSystemContract(
+                contract_name,
+            ));
+        };
```

Comment:
Seems like all this is a waste if the fee-handling is `PayToProposer` or `Burn`.

I'd suggest an early return like:
```rust
match engine_config.fee_handling() {
    FeeHandling::PayToProposer | FeeHandling::Burn => return Ok(()),
    FeeHandling::Accumulate => {}
}
```

Reply:
Fair point, applied. I haven't see this opportunity earlier